### PR TITLE
test: improvements and fixes

### PR DESCRIPTION
**Pull request by evilrobot-01**
Some improvements and fixes whilst reviewing #240.

See commit log for further details.

PS - sincere apologies about the last two commits, which were triggered by the fact that this PR tree has not been rebased on main. Rather than reformatting everything, to reformat again in the coming days, I think it made sense to bring in the nightly formatting and apply the fixes accordingly. We would have to tackle it imminently in any case. You should be able to select a range of commits when reviewing to exclude the last two.

### CODE REVIEW

**Review by peterwht — APPROVED** [review]
Nice clean commits.

**Review by chungquantin — COMMENTED** [review]
Looks clean overall, left a few comments for my concerns.

File: extension/src/decoding.rs [hunk 1]
```diff
@@ -130,52 +117,24 @@ mod tests {
...
-			let mut env = mock_environment(0, input.clone());
-			// Decode `input` to `output` using a default processor.
-			assert_ok!(EnumDecodes::<IdentityProcessor>::decode(&mut env));
-			// Decode charges weight based on the length of the input.
-			assert_eq!(env.charged(), read_from_buffer_weight(input.len() as u32));
-		});
-	}
-
-	#[test]
-	fn decode_with_remove_first_byte_processor_works() {
```
[the remark this answers is not in the document]
Comment (evilrobot-01) [reply extension/src/decoding.rs:157]: Its in the `mock` module only and will not be used in any runtime apart from the mocked one. Sadly it needs to be public due to the requirements imposed by `construct_runtime`, but the mock module is not public and therefore it should not be accessible externally. It is also only conditionally compiled when tests are being run.

If you feel that code for testing only needs its own tests, I will add a unit test for the implementation of `RemoveFirstByte` but I think that is overkill personally.

File: extension/src/lib.rs [hunk 2]
```diff
@@ -104,3 +109,123 @@ impl LogTarget for () {
...
+			let call =
+				RuntimeCall::System(Call::remark_with_event { remark: "pop".as_bytes().to_vec() });
+			let encoded_call = call.encode();
+			let mut env = MockEnvironment::new(DispatchExtFuncId::get(), encoded_call.clone());
+			let mut extension = Extension::<mock::Config>::default();
+			assert!(matches!(extension.call(&mut env), Ok(Converging(0))));
+			// Charges weight.
+			assert_eq!(
+				env.charged(),
+				overhead_weight(encoded_call.len() as u32) +
```
[the remark this answers is not in the document]
Comment (evilrobot-01) [reply extension/src/lib.rs:159]: I personally prefer seeing exactly what is going into the weight for each of the testing scenarios as I can understand the codebase by just looking at the tests. Sticking it behind a reusable function makes it harder for anyone to see this, as you need to navigate to some other function to get an understanding of what the charged weight actually is. Each test case should be reasonaly unique to not need reusable functions.

The test is also sufficiently small to not need further refactoring yet.

File: extension/src/lib.rs [hunk 3]
```diff
@@ -104,3 +109,123 @@ impl LogTarget for () {
...
+	fn dispatch_call_with_invalid_input_returns_error() {
+		// Invalid encoded runtime call.
+		let input = vec![0u8, 99];
+		let mut env = MockEnvironment::new(DispatchExtFuncId::get(), input.clone());
+		let mut extension = Extension::<mock::Config>::default();
+		assert!(extension.call(&mut env).is_err());
+		// Charges weight.
+		assert_eq!(
+			env.charged(),
+			overhead_weight(input.len() as u32) + read_from_buffer_weight(input.len() as u32)
```
[the remark this answers is not in the document]
Comment (evilrobot-01) [reply extension/src/lib.rs:176]: Is it not sufficiently clear to see that this call results in weight being charged for the extension call overhead, and then reading from the contract? For developers of the crate this is way more important than seeing `pre_dispatch() + dispatch() + post_dispatch()` and then needing to look at the actual implementations to then figure out how its being charged. 

I just dont feel such changes are important right now. It is more important that tests prove something to work as simply as possible, with the emphasis instead on high quality code that is actually used.

File: extension/src/functions.rs [hunk 4]
```diff
@@ -206,7 +204,8 @@ impl<Runtime: pallet_contracts::Config> Function for Tuple {
 	fn execute(
 		env: &mut (impl Environment<Config = Self::Config> + BufIn + BufOut),
 	) -> Result<RetVal> {
-		// Attempts to match a specified extension/function identifier to its corresponding function, as configured by the runtime.
+		// Attempts to match a specified extension/function identifier to its corresponding
+		// function, as configured by the runtime.
 		for_tuples!( #(
             if Tuple::matches(env) {
```
[the remark this answers is not in the document]
Comment (evilrobot-01) [reply extension/src/functions.rs:210]: Thank you, will address.

File: extension/src/decoding.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (chungquantin) [reply extension/src/decoding.rs:157]: @evilrobot-01 I see, it is a valid point. Then I don't think we need that, I mistakenly missed the part that `RemoveFirstByte` is imported from `mock.rs`. If so, I think we don't need to cover the test for that.

File: extension/src/functions.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (evilrobot-01) [reply extension/src/functions.rs:210]: Addressed in latest commit.

File: extension/src/lib.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (evilrobot-01) [reply extension/src/lib.rs:176]: If it were non-test code I would agree and gladly oblige, but owing to time I feel it is good enough as is. I dont want to go through another round of reviews for code that isnt going to change the outcome of the tests.

**Review by chungquantin — APPROVED** [review]
LGTM!

**Pull request merged**